coordinator/src/main/scala/filodb.coordinator/DatasetCoordinatorActor.scala

```diff
@@ -0,0 +1,237 @@
+                                 columnStore: ColumnStore,
+                                 reprojector: Reprojector,
+                                 config: Config) extends BaseActor {
+  import DatasetCoordinatorActor._
+  import context.dispatcher
+
+  val datasetName = projection.dataset.name
+  val nameVer = s"$datasetName/$version"
+  val flushTriggerRows = config.getLong("memtable.flush-trigger-rows")
+  val maxRowsPerTable = config.getInt("memtable.max-rows-per-table")
+  val minFreeMb = config.as[Option[Int]]("memtable.min-free-mb").getOrElse(DefaultMinFreeMb)
+
+  def activeRows: Int = activeTable.numRows
+  def flushingRows: Int = flushingTable.map(_.numRows).getOrElse(-1)
+
+  var activeTable: MemTable[K] = makeNewTable
+  var flushingTable: Option[MemTable[K]] = None
+
+  var curReprojection: Option[Future[Seq[String]]] = None
+  var flushesStarted = 0
+  var flushesSucceeded = 0
+  var flushesFailed = 0
+
+  def makeNewTable(): MemTable[K] = {
+    new MapDBMemTable(projection, config)
+  }
+
+  private def reportStats(): Unit = {
+    logger.info(s"MemTable active table rows: $activeRows")
+    logger.info(s"MemTable flushing rows: $flushingRows")
+  }
+
+  var flushedCallbacks: List[ActorRef] = Nil
+
+  private def ingestRows(ackTo: ActorRef, rows: Seq[RowReader], seqNo: Long): Unit = {
+    // TODO: gate on total rows in both active and flushing tables, not just active table?
+    if (activeRows >= maxRowsPerTable) {
+      logger.debug(s"MemTable ($nameVer) is at $activeRows rows, cannot accept writes...")
+      return
+    }
+
+    val freeMB = getRealFreeMb
+    if (freeMB < minFreeMb) {
+      logger.info(s"Only $freeMB MB memory left, cannot accept more writes...")
+      logger.info(s"Active table ($nameVer) has $activeRows rows")
+      sys.runtime.gc()
+      return
+    }
+
+    // MemTable will take callback function and call back to ack rows
+    activeTable.ingestRows(rows) { ackTo ! NodeCoordinatorActor.Ack(seqNo) }
+
+    // Now, check how full the memtable is, and if it needs to be flipped, and a flush started
+    if (shouldFlush) self ! StartFlush()
+  }
+
+  private def canIngest: Boolean = getRealFreeMb >= minFreeMb && activeRows < maxRowsPerTable
+
+  private def getRealFreeMb: Int =
+    ((sys.runtime.maxMemory - (sys.runtime.totalMemory - sys.runtime.freeMemory)) / (1024 * 1024)).toInt
+
+  private def startFlush(): Unit = {
+    logger.info(s"Starting new flush cycle for ($nameVer)...")
+    reportStats()
+    if (curReprojection.isDefined || flushingTable.isDefined) {
+      logger.warn("This should not happen. Not starting flush; $curReprojection; $flushingTable")
+      return
+    }
+    flushingTable = Some(activeTable)
+    activeTable = makeNewTable()
+    val newTaskFuture = reprojector.reproject(flushingTable.get, version)
+    curReprojection = Some(newTaskFuture)
+    flushesStarted += 1
+    newTaskFuture.map { results =>
+      self ! FlushDone(results)
+    }
```

Comment:
Or a for comprehension

for (results <- newTaskFuture) self ! FlushDone(results)